### PR TITLE
Replace references to `slave` with `replica` and add deprecation messaging

### PR DESCRIPTION
**Pull request by craig-day**
This revives #199, but instead of wholesale replacement it keeps existing methods and classes but adds deprecation messaging where possible. Yes, the messaging will be SUPER noisy, but I think that is kind of the point here. 

This also only targets replacing `slave` with `replica`. I will handle `master` -> `primary` in a separate PR using a similar mechanism once this gets settled and working.

I've decided to use ActiveSupport's deprecation messaging since we have it as a dependency, and then our library will behave just like ActiveRecord. I did add an ActiveRecordShards specific class for deprecation which inherits from ActiveSupport. This allows us to set independent behaviors for ActiveRecordShards, which is particularly useful in test when we want to fail for upstream deprecations, but we obviously don't want to fail on our own warnings.

### TODO

- [x] how do we want to handle deprecation messaging
- [ ] do we want any explicit tests around parsing old configuration?

### CODE REVIEW

File: lib/active_record_shards/association_collection_connection_selection.rb [hunk 1]
```diff
@@ -1,12 +1,22 @@
...
   module AssociationCollectionConnectionSelection
-    def on_slave_if(condition)
-      condition ? on_slave : self
+    def on_replica_if(condition)
+      condition ? on_replica : self
     end
 
-    def on_slave_unless(condition)
-      on_slave_if(!condition)
+    def on_slave_if(*args)
```
Comment (bquorning) [inline lib/active_record_shards/association_collection_connection_selection.rb]: Can we just use `alias` for the, well, aliases? So we define `def on_replica_if(condition)` and `alias :on_slave_if, :on_replica_if`. I think that will make it easier to clean up later.

File: lib/active_record_shards/configuration_parser.rb [hunk 2]
```diff
@@ -23,9 +23,14 @@ def explode(conf)
       end
 
       conf.to_a.each do |env_name, env_config|
-        if slave_conf = env_config.delete('slave')
-          expand_child!(env_config, slave_conf)
-          conf["#{env_name}_slave"] = slave_conf
+        if replica_conf = env_config.delete('slave')
+          # TODO
+          env_config['replica'] = replica_conf
```
Comment (bquorning) [inline lib/active_record_shards/configuration_parser.rb]: I think this should be the other way around. *If* anyone adds both `slave` and `replica` configuration (while migrating from the old format to the new format) I think we should let the new config (replica) take precedence.

File: lib/active_record_shards/connection_switcher.rb [hunk 3]
```diff
@@ -70,63 +80,78 @@ def on_master_unless(condition, &block)
...
+      on_master_or_replica(:master, &block)
     end
 
     # just to ease the transition from replica to active_record_shards
-    alias_method :with_slave, :on_slave
-    alias_method :with_slave_if, :on_slave_if
-    alias_method :with_slave_unless, :on_slave_unless
+    alias_method :with_replica, :on_replica
+    alias_method :with_replica_if, :on_replica_if
+    alias_method :with_replica_unless, :on_replica_unless
```
Comment (bquorning) [inline lib/active_record_shards/connection_switcher.rb]: No need to add three new aliases. And we need to keep the old ones. Or delete them: Nobody is still transitioning from the `replica` gem to `active_record_shards`.

File: lib/active_record_shards/connection_switcher.rb [hunk 4]
```diff
@@ -152,7 +177,12 @@ def shard_names
     def switch_connection(options)
       if options.any?
         if options.key?(:slave)
-          current_shard_selection.on_slave = options[:slave]
+          # TODO
+          options[:replica] = options.delete(:slave)
```
Comment (bquorning) [inline lib/active_record_shards/connection_switcher.rb]: Here, too, I think we should let `:replica` take precedence over `:slave`.

File: lib/active_record_shards/tasks.rb [hunk 5]
```diff
@@ -9,7 +9,7 @@
   desc 'Drops the database for the current RAILS_ENV including shards'
   task drop: :load_config do
     ActiveRecord::Base.configurations.each do |key, conf|
-      next if !key.starts_with?(ActiveRecordShards.rails_env) || key.ends_with?("_slave")
+      next if !key.starts_with?(ActiveRecordShards.rails_env) || key.ends_with?("_slave") || key.ends_with?("_replica")
```
Comment (bquorning) [inline lib/active_record_shards/tasks.rb]: Nitpick: `… key.ends_with?("_slave", "_replica")`

File: lib/active_record_shards/tasks.rb [hunk 6]
```diff
@@ -29,7 +29,7 @@
   desc "Create the database defined in config/database.yml for the current RAILS_ENV including shards"
   task create: :load_config do
     ActiveRecord::Base.configurations.each do |key, conf|
-      next if !key.starts_with?(ActiveRecordShards.rails_env) || key.ends_with?("_slave")
+      next if !key.starts_with?(ActiveRecordShards.rails_env) || key.ends_with?("_slave") || key.ends_with?("_replica")
```
Comment (bquorning) [inline lib/active_record_shards/tasks.rb]: same here

**Comment by bquorning** [comment]
By the way, you should feel free to add a `.github/term-check.yaml` file with these contents:
```yaml
ignore:
  - lib/active_record_shards/association_collection_connection_selection.rb
  - lib/active_record_shards/configuration_parser.rb
  - lib/active_record_shards/default_slave_patches.rb
  - lib/active_record_shards/model.rb
  - lib/active_record_shards/shard_selection.rb
  - lib/active_record_shards/tasks.rb
```
We know words are there, and we need to keep them for backwards compatibility.

One other option would be to move all backwards compatibility code into one file. I just think it would make the code harder to read *and* potentially lead to bugs further down the road.

File: lib/active_record_shards/association_collection_connection_selection.rb (hunk 1 above)
Comment (pschambacher) [reply lib/active_record_shards/association_collection_connection_selection.rb]: I think that the goal here is to add a deprecation warning where the `# TODO` is.

File: lib/active_record_shards/configuration_parser.rb (hunk 2 above)
Comment (pschambacher) [reply lib/active_record_shards/configuration_parser.rb]: I agree with Benjamin here. We should give precedence to the `replica` key.

**Comment by craig-day** [comment]
This is all good to go, but the rails 3.2 test fails because in 3.2, `ActiveSupport::Deprecation` was a module and not a class. After #197 is merged, this can be rebased and the build should be green.

**Comment by mariozaizar** [comment]
https://github.com/zendesk/active_record_shards/pull/197 has been merged 👀

**Comment by craig-day** [comment]
Rebased from master, conflicts resolved. Ready for re-review!

**Comment by mriddle** [comment]
@craig-day if you've got time, could you look at rebasing this and fixing the conflicts? I'll look at trying to get this in.

If you don't have time and would like someone to take over I'd be happy to.

**Comment by craig-day** [comment]
@mriddle @bquorning this is rebased, updated, and green. Any objections to me merging?

**Comment by craig-day** [comment]
I'm going to merge this to ensure no more conflicts arise, but I will not tag or release it so we can revert it out and not pollute releases.

File: lib/active_record_shards/default_replica_patches.rb [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    end
+    alias_method :columns_with_force_slave, :columns_with_force_replica
+
+    def table_exists_with_force_replica?(*args, &block)
+      on_cx_switch_block(:replica, construct_ro_scope: false, force: true) do
+        table_exists_without_force_replica?(*args, &block)
+      end
+    end
+    alias_method :table_exists_with_force_slave?, :table_exists_with_force_replica?
+
```
Comment (bquorning) [inline lib/active_record_shards/default_replica_patches.rb:53]: These two methods were removed in 790fb4754ebba5370fe60cc52f04d57b00bf5f37, but were accidentally added back in this PR.

This mistake will be remedied in #254, if not sooner.

**Pull request merged**